runtimes/php-8.1/server.php

```diff
@@ -3,96 +3,191 @@
 
 const USER_CODE_PATH = '/usr/code-start';
 
-function join_paths() {
-    $paths = array();
+class Response {
+    function send(string $body, int $statusCode = 200, array $headers = []): array {
+        return [
+            'body' => $body,
+            'statusCode' => $statusCode,
+            'headers' => $headers,
+        ];
+    }
+
+    function json(array $obj, int $statusCode = 200, array $headers = []) {
+        $headers['content-type'] = 'application/json';
+        return $this->send(\json_encode($obj), $statusCode, $headers);
+    }
 
-    foreach (func_get_args() as $arg) {
-        if ($arg !== '') { $paths[] = $arg; }
+    function empty() {
+        return $this->send('', 204, []);
     }
 
-    return preg_replace('#/+#','/',join('/', $paths));
+    function redirect(string $url, int $statusCode = 301, array $headers = []) {
+        $headers['location'] = $url;
+        return $this->send('', $statusCode, $headers);
+    }
 }
 
-class Response {
-    private mixed $response;
-    private int $status = 200;
-    private array $headers = [];
+class Request {
+    public string $bodyString = '';
+    public mixed $body = '';
+    public array $headers = [];
+    public string $method = '';
+    public string $url = '';
+    public string $path = '';
+    public int $port = 80;
+    public string $host = '';
+    public string $scheme = '';
+    public string $queryString = '';
+    public array $query = [];
+}
 
-    function getStatus(): int {
-        return $this->status;
-    }
+class Context {
+    public Request $req;
+    public Response $res;
 
-    function getResponse(): mixed  {
-        return $this->response;
-    }
+    public array $logs = [];
+    public array $errors = [];
 
-    function getHeaders(): array {
-        return $this->headers;
+    function __construct() {
+        $this->req = new Request();
+        $this->res = new Response();
     }
 
-    function send($text, $status = 200) {
-        $this->response = $text;
-        $this->status = $status;
+    function log(mixed $message) {
+        if(\is_array($message)) {
+            $this->logs[] = \json_encode($message);
+        } else {
+            $this->logs[] = \strval($message);
+        }
     }
 
-    function json($json, $status = 200) {
-        $this->status = $status;
-        $this->response = $json;
-        $this->headers['Content-Type'] = 'application/json';
+    function error(mixed $message) {
+        if(\is_array($message)) {
```

Comment:
Do we want to check for `is_object` here too?

Reply:
Agree. added.